stdlib/public/core/NormalizedCodeUnitIterator.swift

```diff
@@ -0,0 +1,253 @@
+//===--- StringNormalization.swift ----------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+internal
+struct NormalizedCodeUnitIterator: IteratorProtocol {
+  var segmentBuffer = _FixedArray16<CodeUnit>(allZeros:())
+  var overflowBuffer: [CodeUnit]? = nil
+  var normalizationBuffer: [CodeUnit]? = nil
+  var source: SegmentSource
+  var segmentBufferIndex = 0
+  var segmentBufferCount = 0
+  var overflowBufferIndex = 0
+  var overflowBufferCount = 0
```

Comment:
I think this is ok for now. I hope to build up more low-level structures and facilities to do this book-keeping for us without significant overhead.